doc/rst/language/spec/bytes.rst

```diff
@@ -0,0 +1,94 @@
+
+.. default-domain:: chpl
+
+.. _Chapter-Bytes:
+
+=====
+Bytes
+=====
+
+The ``bytes`` type is similar to a string but allows arbitrary
+data to be stored in it. Methods on ``bytes`` that interpret the data as
+characters assume that the bytes are ASCII characters.
+
+.. _Bytes_Instantiation:
+
+Bytes Instantiation and Casting
+-------------------------------
+
+A ``bytes`` instance can be created using the literals similar to strings,
+prepended by a `b` character:
+
+.. code-block:: chapel
+
+   var b = b"my bytes";
+
+``bytes`` can also be crated using a specific buffer (i.e. data
+in another *bytes*, a `c_string` or a C pointer) you can use the
+factory functions shown below, such as *createBytesWithNewBuffer*.
+
+.. _Bytes_and_String:
+
+*bytes* and *string*
+~~~~~~~~~~~~~~~~~~~~
+
+As ``bytes`` can store arbitrary data, any :type:`~String.string` can
+be cast to ``bytes``. In that event, the bytes will store UTF-8 encoded
+character data. However, in general, a *bytes* can contain non-UTF-8 bytes
+and needs to be decoded to be converted to a *string*.
+
+.. code-block:: chapel
+
+  var s = "my string";
+  var b = s:bytes;  // this is legal
+
+  /*
+   The reverse is not. The following is a compiler error:
+
+   var s2 = b:string;
+  */
+
+  var s2 = b.decode(); // you need to decode a bytes to convert it to a string
+
+See the documentation for the ``bytes.decode`` method for details.
+
+Similarly, a ``bytes`` can be initialized using a string:
+
+.. code-block:: chapel
+
+   var s = "my string";
+   var b: bytes = s;
+
+Casts from *bytes* to a Numeric Type
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This module supports casts from ``bytes`` to numeric types. Such
+casts will interpret the *bytes* as ASCII characters and convert it
+to the numeric type and throw an error if the *bytes* does not
+match the expected format of a number. For example:
+
+.. code-block:: chapel
+
+  var b = b"a";
+  var number = b:int;
+
+throws an error when it is executed, but
+
+.. code-block:: chapel
+
+  var b = b"1";
+  var number = b:int;
+
+stores the value ``1`` in ``number``.
+
+To learn more about handling these errors, see the
+:ref:`Error Handling technical note <readme-errorHandling>`.
```

Comment:
This reference should also point to the error handling spec section first (as called out on the other PR)